line_profiler/_line_profiler.pyx

```diff
@@ -128,9 +145,39 @@ cdef class LineProfiler:
             import warnings
             warnings.warn("Could not extract a code object for the object %r" % (func,))
             return
-        if code not in self.code_map:
-            self.code_map[code] = {}
-            self.functions.append(func)
+
+        if code.co_code in self.dupes_map:
+            self.dupes_map[code.co_code] += [code]
+            # code hash already exists, so there must be a duplicate function. add no-op
+            co_code = code.co_code + (9).to_bytes(1, byteorder=byteorder) * (len(self.dupes_map[code.co_code]))
+            CodeType = type(code)
+            
+            if hasattr(code, 'replace'):
+                # python 3.8+
+                code = func.__code__.replace(co_code=co_code)
+            else:
+                # python <3.8
+                co = code
+                code = CodeType(co.co_argcount, co.co_kwonlyargcount,
+                                co.co_nlocals, co.co_stacksize, co.co_flags,
+                                co_code, co.co_consts, co.co_names,
+                                co.co_varnames, co.co_filename, co.co_name,
+                                co.co_firstlineno, co.co_lnotab, co.co_freevars,
+                                co.co_cellvars)
+            func.__code__ = code
+        else:
+            self.dupes_map[code.co_code] = [code]
+        # TODO: Since each line can be many bytecodes, this is kinda inefficient
+        # See if this can be sped up by not needing to iterate over every byte
+        for offset, byte in enumerate(code.co_code):
+            code_hash = hash((code.co_code)) ^ PyCode_Addr2Line(<PyCodeObject*>code, offset)
```

Comment:
This is effectively the same logic, yes, however we might suffer a performance penalty if we have a single method/function that has to handle the case of using PyCode_Addr2Line and the case of just accessing the py_frame.f_lineno member. I believe that'd be a runtime check.

Reply:
`PyCode_Addr2Line` returns a line number, so just send that to the same inline function that accepts two int64. Granted, you are just writing a function that does xor. 

However, from a code maintained and readability perspective it becomes much clearer what is going on. Currently it just reads as a random xor operation. But with a function call it becomes clear that you are combining information, and the way you are doing that is via an xor. Also consider the case where we discover there is a flaw in the way line numbers are combined with the hashes. Adding this abstraction means you only need to update one place instead of two.

Reply:
I see your point, I'll add that now!